library/ssl_cli.c

```diff
@@ -1878,7 +1878,7 @@ static int ssl_parse_hello_verify_request( mbedtls_ssl_context *ssl )
      * even is lower than our min version.
      */
     if( major_ver < MBEDTLS_SSL_MAJOR_VERSION_3 ||
-        minor_ver < MBEDTLS_SSL_MINOR_VERSION_2 ||
+        minor_ver < MBEDTLS_SSL_MAJOR_VERSION_3 - 1 ||
```

Comment:
`minor_ver < MBEDTLS_SSL_MINOR_VERSION_3`

Reply:
My understanding of the comment above this conditional, that:
```C
     * Since the RFC is not clear on this point, accept DTLS 1.0 (TLS 1.1)
     * even is lower than our min version.
```
is that I need to compare the minor with the previous version which so far was used as a `#define` and in this PR has been removed, so I need to use existing code to calculate proper value. Pleas tell if you need me to resolve it in another way.

Reply:
oh. Hmm, I missed this comment. Anyways, we are removing TLS 1.0 and 1.1, so I am not sure this comment still applies. @gilles-peskine-arm could you please have a look?

Reply:
I don't know offhand. @mpg originally wrote that comment so he probably knows better than I do.

Reply:
I think we need to accept DTLS 1.0 (TLS 1.1) here as a compliant peer might send this. Quoting from [RFC 6347 page 16]():

   The server_version field has the same syntax as in TLS.  However, in
   order to avoid the requirement to do version negotiation in the
   initial handshake, DTLS 1.2 server implementations SHOULD use DTLS
   version 1.0 regardless of the version of TLS that is expected to be
   negotiated.

Even if we're only supporting DTLS 1.2, the server might still send DTLS 1.0 in this particular message and still successfully negotiate DTLS 1.2 later, so we absolutely should accept DTLS 1.0 here.

So, I think the code is correct as it stands.

Reply:
Note that this is not merely theoretical: I just checked with wireshark and both OpenSSL and GnuTLS send `0xfeff` (DTLS 1.0) as the version here, even though they successfully complete the handshake using DTLS 1.2 afterwards. (It looks like we're sending DTLS 1.2 though.)

Reply:
In that case, I would suggest to not remove the macro definition at all.

Reply:
Yes, sounds like reasonable suggestion but we should prevent users to use the macros in the new code - maybe just a comment beside the macro definition should be enough?

Reply:
If the definition is in an internal header (under `library` rather than `include/mbedtls`) it should be enough to prevent users from using it by mistake.

Reply:
Definition was in `library/ssl_misc.h `.
@TRodziewicz please restore it.

Reply:
OK, I'm on it.